src/main/java/gov/usgs/earthquake/distribution/WebSocketClient.java

```diff
@@ -10,121 +10,105 @@
 @ClientEndpoint
 public class WebSocketClient implements Runnable {
 
-  public static final long DEFAULT_TIMEOUT_MILLIS = 30000;
+  public static final long DEFAULT_RECONNECT_INTERVAL = 30000;
 
   private Session session;
   private URI endpoint;
   private WebSocketListener listener;
   private Thread thread;
-  private long timeoutMillis;
+  private long reconnectInterval;
   private boolean stopThread;
-  private boolean doReconnect;
-  private Object sync = new Object();
+  private boolean reconnect;
+  private Object close = new Object();
 
   public WebSocketClient (URI endpoint, WebSocketListener listener) {
-    this (endpoint, listener, DEFAULT_TIMEOUT_MILLIS);
+    this (endpoint, listener, true, DEFAULT_RECONNECT_INTERVAL);
   }
 
-  public WebSocketClient(URI endpoint, WebSocketListener listener, long timeoutMillis) {
+  public WebSocketClient(URI endpoint, WebSocketListener listener, boolean reconnect, long reconnectInterval) {
     this.listener = listener;
     this.endpoint = endpoint;
-    this.timeoutMillis = timeoutMillis;
+    this.reconnect = reconnect;
+    this.reconnectInterval = reconnectInterval;
   }
 
 
   public void startup() {
     stopThread = false;
-    this.thread = new Thread(this);
-    this.thread.start();
+    thread = new Thread(this);
+    thread.start();
   }
 
   public void shutdown() throws Exception {
     // bring down thread
     stopThread = true;
-    thread.interrupt();
+    synchronized (close) {
+      thread.interrupt();
+    }
     thread.join();
+    thread = null;
+  }
+
+  public void connect() throws Exception {
+    // try to connect to server
+    WebSocketContainer container = ContainerProvider.getWebSocketContainer();
+    container.connectToServer(this, endpoint);
+  }
+
+  public void disconnect() throws Exception {
+    if (isConnected()) {
+      session.close();
+    }
   }
 
   public void run() {
-    synchronized (sync) {
-      while (!stopThread) {
-        doReconnect = false;
-        try {
+    while (!stopThread) {
+      // try to connect
+      try {
+        synchronized(close) {
           connect();
-          if (!doReconnect) {
-            sync.wait();
-          }
-        } catch (InterruptedException e1) {
-          // likely told to stop
-          stopThread = true;
-        } catch (Exception e2) {
-          // unresolvable problem connecting; try to resolve with listener
           try {
-            listener.onConnectException(e2);
-          } catch (Exception e3) {
-            // unable to handle; quit
+            // successful connection
+            close.wait();
+          } catch (InterruptedException ie) {
+            // gracefully close, disconnect if possible
             stopThread = true;
```

Comment:
`stopThread` is for other threads to communicate in, and this shouldn't change it.
Move the call to `disconnect` to  outside the stopThread loop, to cleanup when the thread is exiting.

Move the InterruptedException catch to the outer try, nothing to be done other than skip call to onConnectException and reconnect sleep.

Reply:
removed all stopThread modifications in run(); threw InterruptedException to be handled by outer try/catch.